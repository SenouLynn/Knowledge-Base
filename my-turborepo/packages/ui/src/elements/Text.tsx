import createFactoryElement from "../sematic/factory";

type TextProps = {
  type: "mainHeader" | "header" | "subHeader" | "body" | "caption";
  as: "p" | "span" | "h1" | "h2" | "h3" | "h4" | "h5" | "h6";
};
