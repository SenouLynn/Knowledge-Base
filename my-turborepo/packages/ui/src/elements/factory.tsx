import React from "react";

const createFactoryElement = <T,>(tagName: string): Element<T> => {
  return (props) => {
    return React.createElement(tagName, props, props.children);
  };
};

/* Enforces types passed to it */
export const createStrictElement = <V,>(
  tagName: string
): Element<ElementPropsFactory<"strict", V>> => {
  return createFactoryElement(tagName);
};

/* Does not enforce types */
export const createLooseElement = <V,>(
  tagName: string
): Element<ElementPropsFactory<"loose", V>> => {
  return createFactoryElement(tagName);
};

export default createFactoryElement;

type GenericElementProps = {
  children?: React.ReactNode;
  // Add any additional props specific to the HTML element here
};

type Element<ComponentProps> = (
  props: ComponentProps & GenericElementProps
) => JSX.Element;

type ElementPropsFactory<T extends "strict" | "loose", V> = T extends "strict"
  ? V
  : any;
