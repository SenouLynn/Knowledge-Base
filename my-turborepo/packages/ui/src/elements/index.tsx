import createFactoryElement, { createStrictElement } from "./factory";

const factoryElement = <Types,>({ tagName }: BuildElement) =>
  createStrictElement<Types>(tagName);

type ContainerProps = {
  testarooney: "test";
};
const elements = {
  div: factoryElement<ContainerProps>({ tagName: "div" }),
  span: factoryElement({ tagName: "span" }),
  h1: factoryElement({ tagName: "h1" }),
  h2: factoryElement({ tagName: "h2" }),
  h3: factoryElement({ tagName: "h3" }),
  h4: factoryElement({ tagName: "h4" }),
  h5: factoryElement({ tagName: "h5" }),
  h6: factoryElement({ tagName: "h6" }),
  p: factoryElement({ tagName: "p" }),
  a: factoryElement({ tagName: "a" }),
  img: factoryElement({ tagName: "img" }),
  button: factoryElement({ tagName: "button" }),
  input: factoryElement({ tagName: "input" }),
  label: factoryElement({ tagName: "label" }),
  select: factoryElement({ tagName: "select" }),
  textarea: factoryElement({ tagName: "textarea" }),
  form: factoryElement({ tagName: "form" }),
  ul: factoryElement({ tagName: "ul" }),
  li: factoryElement({ tagName: "li" }),
  ol: factoryElement({ tagName: "ol" }),
  table: factoryElement({ tagName: "table" }),
  thead: factoryElement({ tagName: "thead" }),
  tbody: factoryElement({ tagName: "tbody" }),
  tr: factoryElement({ tagName: "tr" }),
  th: factoryElement({ tagName: "th" }),
  td: factoryElement({ tagName: "td" }),
  nav: factoryElement({ tagName: "nav" }),
  header: factoryElement({ tagName: "header" }),
  footer: factoryElement({ tagName: "footer" }),
  aside: factoryElement({ tagName: "aside" }),
  section: factoryElement({ tagName: "section" }),
  article: factoryElement({ tagName: "article" }),
  main: factoryElement({ tagName: "main" }),
  body: factoryElement({ tagName: "body" }),
  html: factoryElement({ tagName: "html" }),
  head: factoryElement({ tagName: "head" }),
  meta: factoryElement({ tagName: "meta" }),
  title: factoryElement({ tagName: "title" }),
  link: factoryElement({ tagName: "link" }),
  script: factoryElement({ tagName: "script" }),
  style: factoryElement({ tagName: "style" }),
  noscript: factoryElement({ tagName: "noscript" }),
  iframe: factoryElement({ tagName: "iframe" }),
  canvas: factoryElement({ tagName: "canvas" }),
  svg: factoryElement({ tagName: "svg" }),
  audio: factoryElement({ tagName: "audio" }),
  video: factoryElement({ tagName: "video" }),
  source: factoryElement({ tagName: "source" }),
  track: factoryElement({ tagName: "track" }),
  embed: factoryElement({ tagName: "embed" }),
  object: factoryElement({ tagName: "object" }),
  map: factoryElement({ tagName: "map" }),
  area: factoryElement({ tagName: "area" }),
  time: factoryElement({ tagName: "time" }),
  code: factoryElement({ tagName: "code" }),
  pre: factoryElement({ tagName: "pre" }),
  blockquote: factoryElement({ tagName: "blockquote" }),
  abbr: factoryElement({ tagName: "abbr" }),
  address: factoryElement({ tagName: "address" }),
  cite: factoryElement({ tagName: "cite" }),
  b: factoryElement({ tagName: "b" }),
  strong: factoryElement({ tagName: "strong" }),
  i: factoryElement({ tagName: "i" }),
  em: factoryElement({ tagName: "em" }),
  small: factoryElement({ tagName: "small" }),
  sub: factoryElement({ tagName: "sub" }),
  sup: factoryElement({ tagName: "sup" }),
  samp: factoryElement({ tagName: "samp" }),
};

type BuildElement = { tagName: string; props?: any; children?: any[] };

export default elements;
