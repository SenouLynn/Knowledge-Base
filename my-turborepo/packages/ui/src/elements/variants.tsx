import elements from "../sematic";

export const TextFactory = ({ tagName }: { tagName: keyof typeof elements }) =>
  elements[tagName];

export const variants = {
  MainHeader: TextFactory({ tagName: "h1" }),
  MainSubHeader: TextFactory({ tagName: "h2" }),
  Header: TextFactory({ tagName: "h3" }),
  SubHeader: TextFactory({ tagName: "h4" }),
  LabelHeader: TextFactory({ tagName: "h5" }),
  LabelSubHeader: TextFactory({ tagName: "h6" }),

  Body: TextFactory({ tagName: "p" }),
  Text: TextFactory({ tagName: "p" }),

  Label: TextFactory({ tagName: "label" }),
};
