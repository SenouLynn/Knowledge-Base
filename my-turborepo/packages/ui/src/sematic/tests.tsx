import tests from "tests";
import elements from ".";

const runElementTest = (element: string) => {
  describe(`Running '${element}' factory tests: `, () => {
    const override = "override";

    const Component = (props: any) =>
      elements[element as keyof typeof elements]({
        "data-testid": element,
        ...props,
        children: props.children,
      });

    it(`'${element}'` + " should render as element", () => {
      const render = () => tests.render(<Component data-testid={element} />);
      expect(render).not.toThrow();
    });

    it(`'${element}'` + " should render as wrapper", () => {
      const stringChild = "child";
      const render = () =>
        tests.render(
          <Component data-testid={element}>{stringChild}</Component>
        );
      expect(render).not.toThrow();
      tests.findText(stringChild);
    });

    it(`'${element}'` + " should render correct html element", () => {
      const render = () => tests.render(<Component />);
      expect(render).not.toThrow();
      tests.findHtmlElement(element);
    });

    it(`'${element}'` + " should expose data-testid for external use", () => {
      const render = () => tests.render(<Component data-testid={override} />);
      expect(render).not.toThrow();
      tests.findHtmlElement(override);
    });

    it(`'${element}'` + "should expose role for external use", () => {
      const render = () => tests.render(<Component role={override} />);
      expect(render).not.toThrow();
      tests.findRole(override);
    });

    it.skip("should expose className for external use", () => {
      const render = () => tests.render(<Component className={override} />);
      expect(render).not.toThrow();
      //   expect(tests.findHtmlElement(override)).toHaveClass(override);
    });
  });
};

export default runElementTest;
