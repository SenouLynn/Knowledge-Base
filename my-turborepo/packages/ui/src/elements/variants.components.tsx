import { buildElement } from "../sematic";

export const createComponentProps = () => {};
export const buildComponent = ({
  element,
  role,
}: {
  element: Function;
  role: string;
}) => {
  return (props: any) => {
    return element({ role, ...props });
  };
};

const Components = {
  MainHeader: buildComponent({
    element: buildElement({ tagName: "h1" }),
    role: "main-header",
  }),
  SubHeader: buildComponent({
    element: buildElement({ tagName: "h2" }),
    role: "sub-header",
  }),
  Header: buildComponent({
    element: buildElement({ tagName: "h3" }),
    role: "header",
  }),
  LabelHeader: buildComponent({
    element: buildElement({ tagName: "h4" }),
    role: "label-header",
  }),
  LabelSubHeader: buildComponent({
    element: buildElement({ tagName: "h5" }),
    role: "label-sub-header",
  }),
  Body: buildComponent({
    element: buildElement({ tagName: "p" }),
    role: "body-text",
  }),
  Text: buildComponent({
    element: buildElement({ tagName: "p" }),
    role: "text",
  }),
  SubText: buildComponent({
    element: buildElement({ tagName: "p" }),
    role: "sub-text",
  }),
  Label: buildComponent({
    element: buildElement({ tagName: "label" }),
    role: "label",
  }),
  FormText: buildComponent({
    element: buildElement({ tagName: "p" }),
    role: "form-text",
  }),
  FormSubText: buildComponent({
    element: buildElement({ tagName: "p" }),
    role: "form-subtext",
  }),
  FormHeader: buildComponent({
    element: buildElement({ tagName: "h2" }),
    role: "form-header",
  }),
  FormSubHeader: buildComponent({
    element: buildElement({ tagName: "h3" }),
    role: "form-subheader",
  }),
  FormFooter: buildComponent({
    element: buildElement({ tagName: "p" }),
    role: "form-footer-text",
  }),
  FormLabel: buildComponent({
    element: buildElement({ tagName: "h5" }),
    role: "form-label",
  }),
  FormSubLabel: buildComponent({
    element: buildElement({ tagName: "h6" }),
    role: "form-sublabel",
  }),
  FormInputLabel: buildComponent({
    element: buildElement({ tagName: "label" }),
    role: "form-input-label",
  }),
  FormInputSubLabel: buildComponent({
    element: buildElement({ tagName: "label" }),
    role: "form-input-sub-label",
  }),
  InputError: buildComponent({
    element: buildElement({ tagName: "p" }),
    role: "form-input-error",
  }),
  InputFooter: buildComponent({
    element: buildElement({ tagName: "p" }),
    role: "form-input-footer",
  }),
};

export default Components;
