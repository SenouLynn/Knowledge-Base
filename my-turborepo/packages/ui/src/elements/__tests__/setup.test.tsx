import tests from "tests";
import elements from "..";

describe("Run sematic element tests", () => {
  tests.happyPath("Sematic element tests: happy path :)");

  tests.simpleTest({
    label: "Happy path pass through :)",
    chkVal: true,
    run: (chkVal) => chkVal,
    check: (res, chkVal) => expect(res).toBe(chkVal),
  });

  tests.simpleTest({
    label: "Sad path pass through :(",
    chkVal: true,
    run: (chkVal) => !chkVal,
    check: (res, chkVal) => expect(res).not.toBe(chkVal),
  });
});

describe("Elements render", () => {
  const { div } = elements;
  const Component = (props: any) =>
    div({ testarooney: "test" as const, children: props.children });

  it("should as element", () => {
    const Div = () => tests.render(<Component />);
    expect(Div).not.toThrow();
  });
  it("should as wrapper", () => {
    const child = <>child</>;
    const Div = () => tests.render(<Component>{child}</Component>);
    expect(Div).not.toThrow();
    expect(screen)
  });
});
