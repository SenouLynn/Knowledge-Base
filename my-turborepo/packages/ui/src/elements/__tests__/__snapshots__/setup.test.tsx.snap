// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elements render should render 1`] = `
<div>
  hello world
</div>
`;
